kaldb/src/test/java/com/slack/kaldb/recovery/RecoveryServiceTest.java

```diff
@@ -164,6 +178,147 @@ public void testShouldHandleRecoveryTask() throws Exception {
+    TestKafkaServer.KafkaComponents components = getKafkaTestServer(S3_MOCK_RULE);
+    KaldbConfigs.KaldbConfig kaldbCfg =
+        makeKaldbConfig(components.testKafkaServer, TEST_S3_BUCKET, topicPartition.topic());
+    metadataStore =
+        ZookeeperMetadataStoreImpl.fromConfig(
+            components.meterRegistry, kaldbCfg.getMetadataStoreConfig().getZookeeperConfig());
+
+    final KaldbKafkaConsumer localTestConsumer =
+        new KaldbKafkaConsumer(
+            topicPartition.topic(),
+            Integer.toString(topicPartition.partition()),
+            components.testKafkaServer.getBroker().getBrokerList().get(),
+            TEST_KAFKA_CLIENT_GROUP,
+            "true",
+            "500",
+            "500",
+            components.logMessageWriter,
+            components.meterRegistry,
+            components.consumerOverrideProps);
+    final Instant startTime =
+        LocalDateTime.of(2020, 10, 1, 10, 10, 0).atZone(ZoneOffset.UTC).toInstant();
+    final long msgsToProduce = 100;
+    TestKafkaServer.produceMessagesToKafka(
+        components.testKafkaServer.getBroker(),
+        startTime,
+        topicPartition.topic(),
+        topicPartition.partition(),
+        (int) msgsToProduce);
+    await().until(() -> localTestConsumer.getEndOffSetForPartition() == msgsToProduce);
+    setRetentionTime(components.adminClient, topicPartition.topic(), 250);
+    with()
+        .atMost(1, TimeUnit.MINUTES)
+        .await()
+        .until(() -> getStartOffset(components.adminClient, topicPartition) > 0);
+
+    // produce some more messages that won't be expired
+    setRetentionTime(components.adminClient, topicPartition.topic(), 25000);
+    TestKafkaServer.produceMessagesToKafka(
+        components.testKafkaServer.getBroker(),
+        startTime,
+        topicPartition.topic(),
+        topicPartition.partition(),
+        (int) msgsToProduce);
+    await()
+        .until(() -> localTestConsumer.getEndOffSetForPartition() == msgsToProduce + msgsToProduce);
+
+    // Start recovery service
+    recoveryService =
+        new RecoveryService(kaldbCfg, metadataStore, components.meterRegistry, blobFs);
+    recoveryService.startAsync();
+    recoveryService.awaitRunning(DEFAULT_START_STOP_DURATION);
+
+    // Start recovery with an offset range that is partially unavailable
+    long startOffset = 50;
+    long endOffset = 150;
+    RecoveryTaskMetadata recoveryTask =
+        new RecoveryTaskMetadata(
+            topicPartition.topic(),
+            Integer.toString(topicPartition.partition()),
+            startOffset,
+            endOffset,
+            Instant.now().toEpochMilli());
+    assertThat(recoveryService.handleRecoveryTask(recoveryTask)).isTrue();
+    assertThat(getCount(RECORDS_NO_LONGER_AVAILABLE, components.meterRegistry)).isEqualTo(50);
+    assertThat(getCount(MESSAGES_RECEIVED_COUNTER, components.meterRegistry)).isEqualTo(51);
```

Comment:
Can you please add the asserts from other tests on failed messages being zero and roll over metrics also?

Reply:
https://github.com/slackhq/kaldb/pull/440/commits/76042b28b430df570f2641d26c9f08c0c833acc1